osu.Game.Rulesets.Mania/Difficulty/ManiaPerformanceCalculator.cs

```diff
@@ -55,30 +58,212 @@ protected override PerformanceAttributes CreatePerformanceAttributes(ScoreInfo s
+                * Math.Pow(p300, (countGreat + 0.5) / totalHits)
+                * Math.Pow(p200, countGood / totalHits)
+                * Math.Pow(p100, countOk / totalHits)
+                * Math.Pow(p50, countMeh / totalHits)
+                * Math.Pow(p0, countMiss / totalHits);
+
+                return -gradient;
+            }
+
+            // Finding the minimum of the function returns the most likely deviation for the hit results.
+            if (isLegacyScore)
+                return FindMinimum.OfScalarFunction(legacyLikelihoodGradient, 30);
+            else
+                return FindMinimum.OfScalarFunction(lazerLikelihoodGradient, 30);
+        }
+
+        private double[] getLegacyJudgements(ScoreInfo score, ManiaDifficultyAttributes attributes)
+        {
+            double[] judgements = new double[5];
+
+            double overallDifficulty = attributes.OverallDifficulty;
+
+            if (attributes.Convert)
+                overallDifficulty = 10;
+
+            double windowMultiplier = 1;
+
+            if (score.Mods.Any(m => m is ModHardRock))
+                windowMultiplier *= 1 / 1.4;
+            else if (score.Mods.Any(m => m is ModEasy))
+                windowMultiplier *= 1.4;
+
+            judgements[0] = Math.Floor(16 * windowMultiplier);
+            judgements[1] = Math.Floor((64 - 3 * overallDifficulty) * windowMultiplier);
+            judgements[2] = Math.Floor((97 - 3 * overallDifficulty) * windowMultiplier);
+            judgements[3] = Math.Floor((127 - 3 * overallDifficulty) * windowMultiplier);
+            judgements[4] = Math.Floor((151 - 3 * overallDifficulty) * windowMultiplier);
+
+            return judgements;
+        }
+
+        private double[] getLazerJudgements(ScoreInfo score, ManiaDifficultyAttributes attributes)
+        {
+            double[] judgements = new double[5];
+
+            var track = new TrackVirtual(10000);
+            score.Mods.OfType<IApplicableToTrack>().ForEach(m => m.ApplyToTrack(track));
+            double clockRate = track.Rate;
+
+            double windowMultiplier = 1 / clockRate;
+
+            if (score.Mods.Any(m => m is ModHardRock))
+                windowMultiplier *= 1 / 1.4;
+            else if (score.Mods.Any(m => m is ModEasy))
+                windowMultiplier *= 1.4;
+
+            if (attributes.OverallDifficulty < 5)
+                judgements[0] = (22.4 - 0.6 * attributes.OverallDifficulty) * windowMultiplier;
+            else
+                judgements[0] = (24.9 - 1.1 * attributes.OverallDifficulty) * windowMultiplier;
+            judgements[1] = (64 - 3 * attributes.OverallDifficulty) * windowMultiplier;
+            judgements[2] = (97 - 3 * attributes.OverallDifficulty) * windowMultiplier;
+            judgements[3] = (127 - 3 * attributes.OverallDifficulty) * windowMultiplier;
+            judgements[4] = (151 - 3 * attributes.OverallDifficulty) * windowMultiplier;
```

Comment:
I don't get how the clockrate is related to the judgements

Reply:
In lazer, the judgement windows for all 4 gamemodes are multiplied by the rate, as outlined in https://github.com/ppy/osu-wiki/pull/8791